BootloaderCommonPkg/Library/SecureBootLib/SecureBootRsa.c

```diff
@@ -39,60 +42,83 @@ DoRsaVerify (
   IN       UINT8            ComponentType,
   IN CONST UINT8           *Signature,
   IN       UINT8           *PubKey,
-  IN       UINT8           *PubKeyHash      OPTIONAL,
+  IN       UINT8           *PubKeyHash,
+  IN       UINT8            PubKeyHashAlg,
+  IN       UINT8            SignKeyType,
+  IN       UINT8            SignHashAlg,
   OUT      UINT8           *OutHash         OPTIONAL
   )
 {
   UINTN            Index;
   RETURN_STATUS    Status;
   UINT8           *TmpPubKey;
   RSA_PUB_KEY     *InpPubKey;
-  UINT8            Digest[SHA256_DIGEST_SIZE];
-  UINT8            PubKeyBuf[RSA_MOD_SIZE + RSA_E_SIZE];
+  UINT8            Digest[HASH_DIGEST_MAX]; //Set max size
+  UINT8            PubKeyBuf[RSA_MOD_SIZE_MAX + RSA_E_SIZE];
+  UINT8            DigestSize;
+  UINT16           RsaModSz=0;
 
   // Verify public key first
   // Hash caculation for key uses different endian
+  if(SignKeyType == SIG_TYPE_RSA2048){
+    RsaModSz = RSA_MOD_SIZE_2048;
+    DEBUG ((DEBUG_INFO, "RSA_MOD_SIZE_2048!\n"));
+  } else if (SignKeyType == SIG_TYPE_RSA3072) {
+    RsaModSz = RSA_MOD_SIZE_3072;
+    DEBUG ((DEBUG_INFO, "RSA_MOD_SIZE_3072!\n"));
+  } else {
+    return RETURN_UNSUPPORTED;
+  }
+
+  DEBUG ((DEBUG_INFO, "SignHashAlg 0x%x!\n", SignHashAlg));
+
   InpPubKey = (RSA_PUB_KEY *)PubKey;
   TmpPubKey = PubKeyBuf;
-  for (Index = 0; Index < RSA_MOD_SIZE; Index++) {
-    TmpPubKey[Index] = InpPubKey->PubKeyData[RSA_MOD_SIZE - 1 - Index];
+  for (Index = 0; Index < RsaModSz; Index++) {
+    TmpPubKey[Index] = InpPubKey->PubKeyData[RsaModSz - 1 - Index];
```

Comment:
Buffer size is RSA_MOD_SIZE_2048. if SignKeyType == SIG_TYPE_RSA3072, out of bound array index.

Reply:
RsaModSz would be set by calling modules based on key info.
Keytype/Auth type to be added to required headers as in config data struct and others,